### PR TITLE
fix tei mngr

### PR DESCRIPTION
**Pull request by RoadRunnr**

### CODE REVIEW

File: src/ergw_tei_mngr.erl [hunk 1]
```diff
@@ -128,16 +111,16 @@ alloc_tei_reply(From, Reply, Name, RndState, State, Data) ->
...
     Data = #data{prefix = Prefix0, prefix_len = PrefixLen} = maybe_update_data(Data0),
     Prefix = Prefix0 bsl (32 - PrefixLen),
     Mask = 16#ffffffff bsr PrefixLen,
     {TEI0, RndState} = rand:uniform_s(16#fffffffe, RndState0),
     TEI = Prefix + (TEI0 band Mask),
-    case ets:insert_new(TID, {KeyFun(TEI), Name}) of
-	true ->
+    case gtp_context_reg:lookup(KeyFun(TEI)) of
+	undefined ->
 	    alloc_tei_reply(From, {ok, TEI}, Name, RndState, State, Data);
```
Comment (vkatsuba) [inline src/ergw_tei_mngr.erl:122]: Do we need here add `ets:insert(gtp_context_reg, {KeyFun(TEI), Name})` as was implemented before in `gtp_context_reg.erl`?

Comment (RoadRunnr) [reply src/ergw_tei_mngr.erl:122]: no, there context will register itself and thereby insert the teid. The is a small window in this we could create colliding TEIDs, but that was there before the change to teid_mngr.

**Comment by mgumz** [comment]
@RoadRunnr does this fix also addresses the gtp-proxy situation?

**Comment by RoadRunnr** [comment]
> gen_server?
> Not gen_statem?

yeah, that was there before. Luckily gen_server and gen_statem calls are interchangeable.

**Comment by RoadRunnr** [comment]
> @RoadRunnr does this fix also addresses the gtp-proxy situation?

yes

**Review by vkatsuba — APPROVED** [review]

**Comment by RoadRunnr** [comment]
> With only allocate_tei/1 it would be nice with a comment why there is no release/deallocate_tei
> Only allocating and never releasing sounds like it would cause resource starvation.
> 
> Or another name, like random_tei/1
> That sounds less like something that needs to be deallocated.

No, the stateless logic will touch this area as well and I don't want to rename it yet again.

**Review by ebengt — APPROVED** [review]

File: src/ergw_gsn_lib.erl [hunk 2]
```diff
@@ -108,6 +111,24 @@ query_usage_report(ChargingKeys, Ctx, PCtx)
...
     session_modification_request(PCtx, IEs, Ctx).
 
+release_gtp_endpoints(Context0) ->
+    Context = release_cntl_endp(Context0),
+    release_data_endp(Context).
+
+release_cntl_endp(#context{control_port = Port, local_control_tei = TEI} = Context)
+  when is_integer(TEI) ->
+    ergw_tei_mngr:release_tei(Port, TEI),
+    Context#context{control_port = undefined, local_control_tei = undefined};
```
[the remark this answers is not in the document]
Comment (RoadRunnr) [reply src/ergw_gsn_lib.erl]: Kind of agree with the readable part.
But I want to set them back to uninitialized state (which is undefined) so I don't have to handle cleared and not-initialized in different ways.

File: src/ergw_tei_mngr.erl [hunk 3]
```diff
@@ -23,6 +23,10 @@
 
 -export([validate_option/1]).
 
+-ifdef(TEST).
+-export([all/0]).
+-endif.
+
 -include("include/ergw.hrl").
 
 -define(SERVER, ?MODULE).
```
[the remark this answers is not in the document]
Comment (RoadRunnr) [reply src/ergw_tei_mngr.erl]: Not really. Even for a gen_statem it's `ServerName` in start_link/4.

File: src/ggsn_gn.erl [hunk 4]
```diff
@@ -592,8 +593,9 @@ encode_eua(Org, Number, IPv4, IPv6) ->
 		      pdp_type_number = Number,
 		      pdp_address = <<IPv4/binary, IPv6/binary >>}.
 
-close_pdp_context(Reason, #{context := Context, pfcp := PCtx, 'Session' := Session}) ->
-    URRs = ergw_gsn_lib:delete_sgi_session(Reason, Context, PCtx),
+close_pdp_context(Reason, #{context := Context0, pfcp := PCtx, 'Session' := Session} = Data) ->
```
[the remark this answers is not in the document]
Comment (RoadRunnr) [reply src/ggsn_gn.erl]: that one and many more. The handler code has matured to a state that it is unlikely that surprising difference in similar functions in GGSN and PGW will pop up.
The only question is, before or after the stateless effort. Needs proper planing.

**Pull request merged**